Scalable jobserver for rustc

Mark-Simulacrum opened the pull request:
This refactors the job queue code for support of [per-rustc process jobservers](https://github.com/rust-lang/rust/pull/67398). Furthermore, it also cleans up the code and refactors the main loop to be more amenable to understanding (splitting into methods and such).

Assignment of tokens to either rustc "threads" or processes is dedicated to the main loop, which proceeds in a strict "least recently requested" fashion among both thread and process token requests. Specifically, we will first allocate tokens to all pending process token requests (i.e., high-level units of work), and then (in per-rustc jobserver mode) follow up by assigning any remaining tokens to rustcs, again in the order that requests came into cargo (first request served first).

It's not quite clear that that model is good (no modeling or so has been done). On the other hand this strategy should mean that long-running crates will get more thread tokens once we bottom out in terms of rustc parallelism than short-running crates, which means that crates like syn which start early on but finish pretty late should hopefully get more parallelism nicely (without any more complex heuristics).

One plausible change that may be worth exploring is making the assignment prefer earlier rustc's, globally, rather than first attempting to spawn new crates and only then increasing parallelism for old crates. syn for example frequently gets compiled in the early storm of dozens of crates so is somewhat unlikely to have parallelism, until fairly late in its compilation.

We also currently conflate under this model the rayon threads and codegen threads. Eventually inside rustc those will probably(?) also be just one thing, and the rustc side of this implementation provides no information as to what the token request is for so we can't do better here yet.

Mark-Simulacrum commented:
CI here looks to be mostly failing because the current branch expects rustc to have -Zjobserver-token-requests.

bors commented:
:umbrella: The latest upstream changes (presumably #7737) made this pull request unmergeable. Please resolve the merge conflicts.

ehuss commented:
Can this get a doc-comment in `job_server.rs` pointing to the documentation in `src/librustc_jobserver/lib.rs` in `rustc`?  It's pretty hard for me to follow in isolation.

Mark-Simulacrum commented:
Hm, do you mean job_queue.rs? job_server.rs doesn't appear to exist in this repository.

I would be happy to provide such a doc comment -- it's worth noting that I'm guessing we'll be landing this behind some sort of -Z flag as well initially (as we test it out and such) just because it's a non-trivial overhaul of the protocol we use with rustc.

ehuss commented:
Oops, yea, typo.  Should be `job_queue.rs`.

Mark-Simulacrum commented:
I've added some (somewhat) high level docs to job_queue and rebased.

alexcrichton commented:
Ok I've read over this and it looks pretty good to me, certainly for the purposes of testing this looks like it'll cover all our bases.

@Mark-Simulacrum do you want to update your PR in rust-lang/rust to have a Cargo submodule with this commit, and then make a try build so we can test that out?

I think longer-term I'd want to refactor the jobserver management in Cargo here, it feels a bit spaghetti-like and I'd like to make it a bit more concise (if we can) to make it a bit more robust. Nothing concrete I can think of, just a general feeling of "this feels bolted on" which is not at all your fault, just something for me to stew on :)

Also, for the vecs, I think they can be relatively trivially converted to hash maps? I'd expect something like:

```rust
let rustcs_with_a_token: HashMap<u32, Vec<Token>>; 
let rustcs_wanting_a_token: HashMap<u32, (usize, Client)>; // `usize` == how many tokens requested
```

Mark-Simulacrum commented:
Yeah, I think the HashMaps would work. I moved between a few different schemes during development and had started out with Vecs.

I'll go ahead and get the rustc branch to point at this branch (and probably rebase both, while I'm at it) and kickoff a try build.

Mark-Simulacrum commented:
Okay I've force pushed this and the rust-lang/rust branch and kicked off a try build over there, hopefully that all works out.

I didn't look at implementing the HashMap stuff as I'm also a bit strapped for time :)

bors commented:
:umbrella: The latest upstream changes (presumably #7776) made this pull request unmergeable. Please resolve the merge conflicts.

Mark-Simulacrum commented:
CI failure is.. scary "E: Failed to fetch https://packages.microsoft.com/ubuntu/16.04/prod/dists/xenial/main/binary-amd64/Packages  Writing more data than expected (622228 > 606754)"

But seems unrelated.

r? @alexcrichton

alexcrichton commented on src/cargo/core/compiler/context/mod.rs:
```diff
@@ -491,4 +491,18 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
+        // Drain the client fully
+        for _ in 0..tokens {
+            while let Err(e) = client.acquire_raw() {
+                log::trace!("failed to acquire jobserver token: {:?}", e);
```
I think perhaps the error should be propagated up here?

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -373,6 +471,19 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
                                 info!("end: {:?}", id);
                                 finished += 1;
+                                while let Some(pos) =
+                                    rustc_tokens.iter().position(|(i, _)| *i == id)
```
As a comment from before, I think we'll want to use a `HashMap` here

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -306,12 +381,31 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
             }
 
+            // If we managed to acquire some extra tokens, send them off to a waiting rustc.
+            let extra_tokens = tokens.len() - (self.active.len() - 1);
```
If you don't mind doing so, I think that we need to refactor this loop within Cargo to be a bit more readable (it's starting to get a bit sprawling).

I'm imagining something like:

```rust
let mut state = RunState::new(self);
loop {
    state.spawn_work_if_possible()?;
    if state.finished() { break; }
    state.grant_rustc_token_requests()?;
    for event in state.wait_for_events()? {
        state.process_event(event)?;
    }
}
```

Basically I'd like to make the structure of the loop much clearer and then have clear documentation for each method what happens where. This would also hopefully lend itself to better-documented fields as well as easily adding debug assertions if necessary.

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -403,9 +514,34 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
+                        to_send_clients.push((id, client));
+                    }
+                    Message::ReleaseToken(id) => {
+                        // Note that this pops of potentially a completely
```
"of potentially" may be stray here?

alexcrichton commented on src/cargo/core/compiler/mod.rs:
```diff
@@ -683,6 +700,9 @@ fn add_error_format_and_color(
     }
     cmd.arg(json);
+    if cx.bcx.config.cli_unstable().jobserver_per_rustc {
+        cmd.arg("-Zjobserver-token-requests");
```
Could this perhaps be joint with the `jobserver_per_rustc` read above? (consolidating the `if` statements)

alexcrichton commented on src/cargo/core/compiler/mod.rs:
```diff
@@ -1081,8 +1101,9 @@ fn on_stderr_line(
     package_id: PackageId,
     target: &Target,
     options: &mut OutputOptions,
+    rustc_client: Option<&Client>,
```
I think it may perhaps be a bit cleaner to instead of passing `rustc_client` around and sending it through messages to instead store it elsewhere. The processes spawned by the job queue I think should all have an optional `Client` associated with them (stored internally in `ProcessBuilder`), and `job_queue` could keep track of a map from id to `Client` in that case.

alexcrichton commented on src/cargo/core/compiler/mod.rs:
```diff
@@ -1210,6 +1232,36 @@ fn on_stderr_line_inner(
 
+    #[derive(serde::Deserialize)]
+    struct JobserverNotification {
+        jobserver_event: Event,
```
I think we'll probably want to bikeshed this a bit, but the json messages coming out of rustc aren't exactly the most structured things in terms of precedent anyway.

I'd ideally prefer to not ignore errors if Cargo is sure that event has to do with jobserver events, whereas right now if there's a slight mismatch Cargo will ignore everything coming from rustc. In any case this is probably a concern best left for stabilization.

Mark-Simulacrum commented:
Okay I've updated the PR, I think I addressed most of your comments.

I opted not to implement the rustc_client out-of-band storage -- that looked pretty complicated; it seems like there's a semi-opaque wall around "Work" / "Job" units and it was unclear to me how to thread state through that... maybe I missed something obvious though.

Furthermore, we eventually do need to keep track of the number of requests for a given JobId, so while we could e.g. do a usize for that, it seems simpler to just send around clients.

Of course if you disagree then happy to try and dig in more :)


I've also -- in the process of refactoring the main loop and extracting things out into methods -- tried to add documentation here and there but since I've been fairly embedded into this code for a few hours now I'm not sure where all confusion lies anymore :) I think someone with a semi-fresh set of eyes (you) would be more helpful in basically saying "this could use a comment" -- happy to write them up then.


The latest set also shifts around the algorithm for jobserver token assignment under the per-rustc model, where before we were eagerly giving tokens to the first rustc that requested them, which plausibly lead to pretty poor "multi-rustc" parallelism in practice (since the first one would likely gobble up tokens, then the second one would do the same, and so forth, presuming highly parallel and quick to request tokens rustcs). I'm not sure we were observing that behavior but it seemed possibly suboptimal.

The new strategy is to dedicate assignment to the main loop, which now proceeds in a strict "least recently requested" fashion among both types of requests. Specifically, we will first allocate tokens to all pending process token requests (i.e., high-level units of work), and then (in per-rustc jobserver mode) follow up by assigning any remaining tokens to rustcs, again in the order that requests came into cargo (first request served first).

It's not quite clear that that model is good -- at least for the latter bit, maybe randomly spreading tokens around would be better, or some other heuristic, but on the other hand this means that it is likely that long-running crates will get more tokens than short-running crates, which means that crates like `syn` which start early on but finish pretty late should hopefully get more parallelism nicely (without any more complex heuristics).

-------

As a fairly orthogonal note I'm also somewhat increasingly unhappy with the jobserver crate's API -- I keep hitting the problem that the "request a token" interface does not provide for cancellation -- and we've also wanted e.g. Acquired::forget in rustc, so I'll probably put some resources there before moving on to fixing up the rustc PR. Generally my plan is to get this landed or in a state ready for landing and then move on to fixing up the rustc PR :)

Mark-Simulacrum commented:
Tests here are all going to fail as I missed a minor detail where previously cargo would always immediately spawn up a rustc process ("sharing" the cargo implicit jobserver token with that rustc) but the new implementation would make that pretty painful -- ideally we'd synthesize the token out of nowhere, but the jobserver crate makes that hard and in any case it's not quite what we want.

I think to get the correct behavior we'd basically end up with a lot of code trying to shuffle around the implicit token -- loosely, it'd look something like what the current codegen stuff does in rustc I imagine (less complicated due to everything being threadsafe though).

I'm going to try and think some more on this, maybe we can discuss synchronously tomorrow too. Don't have any great ideas here :/

Mark-Simulacrum commented:
Oh, also forgot to mention, I updated the PR description with a summary of the PR, though it loosely duplicates what I said already in a comment.

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -36,13 +86,49 @@ pub struct JobQueue<'a, 'cfg> {
+    /// We should only have tokens here "in-between" loops of the
+    /// `drain_the_queue` function, i.e., when we've acquired a bunch of tokens
+    /// and are about to give them to rustcs.
+    unused_tokens: Vec<Acquired>,
```
Oh sorry, to clarify, can this be a separate struct which only lives for the one stack frame where we're running stuff?

alexcrichton replied:
Also, as to this field specifically, I don't think we're going to want it. There should never be any long-term storage for unused tokens.

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -251,26 +366,209 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
+            .len()
+            .saturating_sub(self.pending_queue.len());
+        for _ in 0..tokens_needed {
+            jobserver_helper.request_token();
```
This I'm a little worried about since it will request a huge amount more than before. I don't think we want to do these requests on each turn of the loop, we should still strive to match up a request to an acquire 1:1

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -36,13 +86,49 @@ pub struct JobQueue<'a, 'cfg> {
+    /// If we ever wind up in a situation where an entry in this map is empty,
+    /// then something has gone wrong, as we have a rustc running without tokens
+    /// associated with it.
+    rustc_tokens: HashMap<JobId, Vec<Acquired>>,
```
I think we're going to want to avoid this structure to handle the implicit token, we I believe can get away with the tokens in this map are purely for "extra parallelism" for each rustc. There'd be a separate vector, whose length is less than one from this map, which has all the tokens used to spawn the rustc process in the first place.

Mark-Simulacrum commented:
Okay I dropped the commit which refactored into the unused_tokens and so forth and then restored it a bit.

I also moved the state into a new struct which only exists while we're draining the job queue.

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -251,26 +384,203 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
+                    .release_raw()
+                    .chain_err(|| "failed to release jobserver token")?;
+            }
+        }
```
This might actually be a bit nicer as:

```rust
while !self.to_send_clients.is_empty() && self.active.len() + 1 < self.tokens.len() {
    let token = match self.tokens.pop() { Some(t) => t, None => break; };
    // ...
}
```

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -251,26 +384,203 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
+            if !self.to_send_clients.is_empty() {
+                // remove from the front so we grant the token to the oldest
+                // waiter
+                let (id, client) = self.to_send_clients.remove(0);
```
Since we're pushing and popping onto this, and it's probably modified relatively frequently, this may actually be a relatively good use for a `VecDeque` perhaps?

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -251,26 +384,203 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
+                }
+            }
+        } else {
+            self.tick_progress();
```
I think the purpose of this loop is to actually not call this in the case that the events are not empty, so was this perhaps a stray refactoring error?

Mark-Simulacrum replied:
Especially in the case where rustc is using this channel for the jobserver protocol, events may not be empty for a long time (i.e., we're constantly trying to wait for threads, releasing tokens, etc.). I can drop this from this PR though and file it separately since it's a drive by change if anything.

alexcrichton replied:
Hm ok, mind moving this to a separate commit? I think it'd be best to handle this with a sort of heartbeat message if it becomes problematic. Did you see extraneous stuttering you weren't expecting when building though?

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -251,26 +384,203 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
+                    self.tokens
+                        .push(rustc_tokens.pop().expect("rustc releases token it has"));
+                } else {
+                    panic!("This job does not have tokens associated with it");
```
I think it's probably best to structure this as an `expect`?

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -251,26 +384,203 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
+            Message::ReleaseToken(id) => {
+                // Note that this pops off potentially a completely
+                // different token, but all tokens of the same job are
+                // conceptually the same so that's fine.
```
Could this comment be extended a bit as well? I think this all works because `self.tokens` is largely a "pool" of tokens which we greedily allocate requests from, but we also truncate it before we block on all iterations of the loop. That way we're basically transferring ownership of the token through the `self.tokens` list

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -251,26 +384,203 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
+                            //
+                            // FIXME: this represents a rustc bug: it did not
+                            // release all of its thread tokens but finished
+                            // completely.
```
I think it's fine to remove this fixme, while in the normal course of events this is indeed a bug I think it's fine in general for Cargo to be defensive about this. For example it helps recover when rustc segfaults or things like that.

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -251,26 +384,203 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
+                            // completely.
+                            self.tokens.extend(rustc_tokens);
+                        }
+                        while let Some(pos) =
```
I'm a little worried about this loop, it feels like on larger machines it could cause a lot of shifting which could be problematic. I'll leave some comments elsewhere I think may help with this.

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -128,24 +233,30 @@ impl<'a> JobState<'a> {
+    ///
+    /// This should arrange for the passed client to eventually get a token via
+    /// `client.release_raw()`.
+    pub fn will_acquire(&self, client: &Client) {
```
On passing in the `Client` here, I think it'd still be best to avoid this. I don't think it'll be too hard though with some small tweaks perhaps? Can you try doing something like:

* Add a `Map<Unit, Client>` to `Context`
* Populate this while preparing rustc invocations (only if the jobserver-per-rustc flag is set)
* Internally here we have a map of `Unit` to `JobId`
* This means that given a `JobId` that needs a token, we can get a `Unit`, which we can then use to look up a `Client`

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -27,22 +78,70 @@ use crate::util::{internal, profile, CargoResult, CargoResultExt, ProcessBuilder
+    /// may be present many times (if it has requested multiple tokens).
+    // We use a vec here as we don't want to order randomly which rustc we give
+    // tokens to.
+    to_send_clients: Vec<(JobId, Client)>,
```
I think the ideal data structure that we want here is some sort of "priority map", although I'm not sure that this exists. Given my comments below I think we can reduce this to `Vec<JobId>`, but I think that would be best represented as `HashMap<JobId, u32>` instead to make removal more efficient. 

I think this is generally fine for now, though, and we can use `map.iter().next()` to get the next rustc to allocate a token. We can always figure out a better prioritization system later for allocating tokens to rustc's, but I'd like to make sure we're not doing a bunch of accidental O(n^2) work in Cargo here.

Mark-Simulacrum replied:
How would you feel about a `BTreeMap<JobId, Client>`? That means the oldest (i.e., longest running) jobs get the tokens first?

Mark-Simulacrum replied:
I guess given the other refactor suggested this could also just be a `BinaryHeap<JobId>`, right?

alexcrichton replied:
Oh yeah actually a `BTreeMap` which means oldest sounds pretty reasonable. I don't think we want `BinaryHeap` as-is, (we'd want a custom comparison function at least) and I'm not sure if that supports arbitrary deletion. In any case let's stick to `BTreeMap` for

Mark-Simulacrum commented:
Pushed up a series of commits which fixes most of your comments -- replied to the one I think is not actually necessary to fix.

alexcrichton commented on src/cargo/core/compiler/context/mod.rs:
```diff
@@ -69,6 +69,8 @@ pub struct Context<'a, 'cfg> {
     /// when `pipelining` above is enabled.
     rmeta_required: HashSet<Unit<'a>>,
+
+    pub rustc_clients: HashMap<Unit<'a>, Client>,
```
Mind throwing a comment on this field about what it holds?

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -284,129 +616,33 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
-                self.run(&unit, job, cx, scope)?;
-            }
+            info!(
+                "tokens: {}, rustc_tokens: {}, waiting_rustcs: {}",
```
Should this perhaps get printed just before blocking? It seems sort of weird to print out this information between spawning/granting?

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -284,129 +616,33 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
-                            acquired_token.chain_err(|| "failed to acquire jobserver token")?,
-                        );
-                    }
+            self.tokens.truncate(self.active.len() - 1);
```
Could this get moved right before the actual blocking? That way we wouldn't accidentally throw away tokens while we're processing a bunch of messages.

Mark-Simulacrum commented:
I think that this latest push resolves your comments (dropped `tick_progress` change entirely, will file a follow-up PR once this lands).

I also realize this has a ton of commits now -- do you want me to squash them down?

I'll look at the CI failures later.

Mark-Simulacrum commented:
CI is now passing.

alexcrichton commented:
@bors: r+

Very nice, thanks so much! Commits are fine :)

bors commented:
:pushpin: Commit 7e3e1b128c21803e6e29c6eaedd2070c2e92ee2e has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 7e3e1b128c21803e6e29c6eaedd2070c2e92ee2e -->
<!-- homu: {"type":"Approved","sha":"7e3e1b128c21803e6e29c6eaedd2070c2e92ee2e","approver":"alexcrichton"} -->

bors commented:
:hourglass: Testing commit 7e3e1b128c21803e6e29c6eaedd2070c2e92ee2e with merge 8ee3d927e6877c1e23d8297fdda5f69846efe2bb...
<!-- homu: {"type":"BuildStarted","head_sha":"7e3e1b128c21803e6e29c6eaedd2070c2e92ee2e","merge_sha":"8ee3d927e6877c1e23d8297fdda5f69846efe2bb"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/473adcf1-6656-4c7e-9337-c393e7e0431c/_build/results?buildId=19123)
Approved by: alexcrichton
Pushing 8ee3d927e6877c1e23d8297fdda5f69846efe2bb to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"alexcrichton","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/473adcf1-6656-4c7e-9337-c393e7e0431c/_build/results?buildId=19123"},"merge_sha":"8ee3d927e6877c1e23d8297fdda5f69846efe2bb"} -->

ehuss commented:
This PR seems to be causing panics with debug builds of Cargo.  @Mark-Simulacrum can you check?  The repro I have is:

1. `cargo new foo`
2. `cd foo`
3. `cargo add bitflags`
4. `cargo test`

It panics at [progress.rs:259](https://github.com/rust-lang/cargo/blame/b68b0978ab8012f871c80736fb910d14b89c4498/src/cargo/util/progress.rs#L259) where in this case that function is called with `cur=2` and `max=1`.  Here's the full backtrace:

<details><summary>bactrace</summary>
<pre>
   Compiling bitflags v1.2.1
thread 'main' panicked at 'attempt to subtract with overflow', src/cargo/util/progress.rs:259:21
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:77
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1057
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1426
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:195
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:215
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:463
  11: rust_begin_unwind
             at src/libstd/panicking.rs:371
  12: std::panicking::begin_panic
  13: std::panicking::begin_panic
  14: cargo::util::progress::Format::progress
             at src/cargo/util/progress.rs:259
  15: cargo::util::progress::State::tick
             at src/cargo/util/progress.rs:176
  16: cargo::util::progress::Progress::tick_now
             at src/cargo/util/progress.rs:107
  17: cargo::core::compiler::job_queue::DrainState::tick_progress
             at src/cargo/core/compiler/job_queue.rs:714
  18: cargo::core::compiler::job_queue::DrainState::wait_for_events
             at src/cargo/core/compiler/job_queue.rs:600
  19: cargo::core::compiler::job_queue::DrainState::drain_the_queue
             at src/cargo/core/compiler/job_queue.rs:648
  20: cargo::core::compiler::job_queue::JobQueue::execute::{{closure}}
             at src/cargo/core/compiler/job_queue.rs:384
  21: crossbeam_utils::thread::scope::{{closure}}
             at /Users/eric/.cargo/registry/src/github.com-1ecc6299db9ec823/crossbeam-utils-0.7.0/src/thread.rs:161
  22: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/85976442558bf2d09cec3aa49c9c9ba86fb15c1f/src/libstd/panic.rs:318
  23: std::panicking::try::do_call
             at /rustc/85976442558bf2d09cec3aa49c9c9ba86fb15c1f/src/libstd/panicking.rs:296
  24: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:79
  25: std::panicking::try
             at /rustc/85976442558bf2d09cec3aa49c9c9ba86fb15c1f/src/libstd/panicking.rs:272
  26: std::panic::catch_unwind
             at /rustc/85976442558bf2d09cec3aa49c9c9ba86fb15c1f/src/libstd/panic.rs:394
  27: crossbeam_utils::thread::scope
             at /Users/eric/.cargo/registry/src/github.com-1ecc6299db9ec823/crossbeam-utils-0.7.0/src/thread.rs:161
  28: cargo::core::compiler::job_queue::JobQueue::execute
             at src/cargo/core/compiler/job_queue.rs:384
  29: cargo::util::config::clippy_driver::{{closure}}
  30: cargo::ops::cargo_compile::compile_ws
             at src/cargo/ops/cargo_compile.rs:469
  31: cargo::ops::cargo_compile::compile_with_exec
             at src/cargo/ops/cargo_compile.rs:259
  32: cargo::ops::cargo_compile::compile
             at src/cargo/ops/cargo_compile.rs:248
  33: cargo::ops::cargo_test::compile_tests
             at src/cargo/ops/cargo_test.rs:69
  34: cargo::ops::cargo_test::run_tests
             at src/cargo/ops/cargo_test.rs:21
  35: cargo::commands::test::exec
             at src/bin/cargo/commands/test.rs:162
  36: cargo::cli::execute_subcommand
             at src/bin/cargo/cli.rs:197
  37: cargo::cli::main
             at src/bin/cargo/cli.rs:107
  38: cargo::main
             at src/bin/cargo/main.rs:39
  39: std::rt::lang_start::{{closure}}
             at /rustc/85976442558bf2d09cec3aa49c9c9ba86fb15c1f/src/libstd/rt.rs:67
  40: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:52
  41: std::panicking::try::do_call
             at src/libstd/panicking.rs:296
  42: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:79
  43: std::panicking::try
             at src/libstd/panicking.rs:272
  44: std::panic::catch_unwind
             at src/libstd/panic.rs:394
  45: std::rt::lang_start_internal
             at src/libstd/rt.rs:51
  46: std::rt::lang_start
             at /rustc/85976442558bf2d09cec3aa49c9c9ba86fb15c1f/src/libstd/rt.rs:67
  47: cargo::init_git_transports
</pre>
</details>

Mark-Simulacrum commented:
Hm, not impossible, will take a look. I guess we're generating too many completed jobs?

It'll probably be tomorrow at the earliest though, if necessary please do revert.

Mark-Simulacrum commented:
Filed #7829 to resolve this bug.

ehuss commented on src/cargo/core/compiler/context/mod.rs:
```diff
@@ -491,4 +497,23 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
+                    tokens,
+                    e,
+                );
+            }
```
What is the intent of this loop?  Should it maybe be something like this?

```rust
client.acquire_raw().chain_err(|| {
    format!(
        "failed to fully drain {}/{} token from jobserver at startup",
        i, tokens
    )
})?;
```

Mark-Simulacrum replied:
Hm, I think so -- it doesn't really make any difference (bail! returns from the method anyway), but I'll send a PR cleaning it up. I suspect I had intended this as "loop until you error" but that's clearly not what this does :)
